### PR TITLE
Make the perfherder graph width sized dynamically, and refine the tooltip

### PR DESCRIPTION
**Pull request by gregtatum**
This PR changes some of the visual layout behavior of the perfherder graph. This change was motivated with the inconsistent feeling of looking at the data on different monitor setups. I frequently move between smaller screens, larger ones, and split views. The graphs were frequently not visible.

#6861 needs to land first, and then this needs to rebase through it, as they are touching similar parts of the code.

Each commit contains a message explaining more in detail what the changes are doing, so I would recommend looking at each commit individually with the message. Here are some screenshots detailing the changes.

# Commit 1 - Chart width

### Large width:

Before, the view was maxed out at 1350px.
> image

After, the graph will take all available width.
> image

### Small width:

Before, the view scaled the entire chart down, including labels. For me, the labels were frequently too small to read. The tooltip also scaled according to the size of the chart, moving outside of the range of legibility.
> image

After, the labels are always consistently sized, and the font doesn't change size.
> image

# Commit 2 - Tooltip

Normally, I would break something like this into a separate PR, but commit 1 regressed the tooltip behavior, so that it would render off the side of the screen. I needed to fix this issue, and as I did, I included some usability fixes as well to make the tooltips more consistent when mousing over the graph.

### Typical view:

image

### Left overflow:
image

### Right overflow:
image

### Fixed vertical positioning

Before, it was done through constant values, which would be weird for certain content.
image

After, the divs can dynamically grow up.
image

### CODE REVIEW

File: ui/perfherder/graphs/GraphTooltip.jsx [hunk 1]
```diff
@@ -138,152 +140,175 @@ const GraphTooltip = ({
...
-    y: y - (186 + verticalOffset),
-  };
+  // Ensure the tooltip does not leave the graph.
+  let tooltipLeft = x - TOOLTIP_WIDTH / 2;
+  const tooltipRight = x + TOOLTIP_WIDTH / 2;
+  let arrowOffset = 0;
+  if (tooltipRight > graphWidth) {
+    // The tooltip is off the right of the container.
+    arrowOffset = tooltipRight - graphWidth;
+    tooltipLeft -= arrowOffset;
```
Comment (alexandru-io) [inline ui/perfherder/graphs/GraphTooltip.jsx:150]: Why are you decrementing with `arrowOffset` while it's set to 0?

File: ui/perfherder/graphs/GraphTooltip.jsx [hunk 2]
```diff
@@ -17,6 +17,8 @@ import RepositoryModel from '../../models/repository';
 import { displayNumber, getStatus } from '../helpers';
 import Clipboard from '../../shared/Clipboard';
 
+const TOOLTIP_WIDTH = 280;
```
Comment (sarah-clements) [inline ui/perfherder/graphs/GraphTooltip.jsx:20]: I think this should be moved inside the component to line 142 or so.

File: ui/perfherder/graphs/GraphTooltip.jsx [hunk 3]
```diff
@@ -138,152 +140,175 @@ const GraphTooltip = ({
...
+    arrowOffset = tooltipRight - graphWidth;
+    tooltipLeft -= arrowOffset;
+  }
+  if (tooltipLeft < 0) {
+    // The tooltip is off the left of the container.
+    arrowOffset = tooltipLeft;
+    tooltipLeft = 0;
+  }
+
+  // const tooltipTop = y - 201;
```
Comment (sarah-clements) [inline ui/perfherder/graphs/GraphTooltip.jsx:158]: Remove comment

File: ui/shared/WithSize.jsx [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        'visibilitychange',
+        this.visibilityChangeListener,
+      );
+
+      // Wrapping the first update in a requestAnimationFrame to defer the
+      // calculation until the full render is done.
+      requestAnimationFrame(() => {
+        // This component could have already been unmounted, check for the existence
+        // of the container.
+        if (this._container) {
```
Comment (sarah-clements) [inline ui/shared/WithSize.jsx:59]: Redundant comment

**Review by sarah-clements — COMMENTED** [review]
These look like nice improvements. If the sheriffs are ok with these changes then that works for me.

**Comment by alexandru-io** [comment]
The only thing I found not ok while testing this PR is:
1. Go to perfherder Graph view
2. Click on `Add test data` button
3. Change the repo filter from `mozilla-central` to `autoland` (to get any test so you can select one)
4. Select first test (basically any test) and click `Plot graphs`
5. You'll have your graphs plotted **but** the `Add test data` modal appears again like you clicked on something like `Plot graphs and add more tests`. Like I suggested at # 4 above, this should happen with any filter you use and any number of tests you add
Screenshot from 2020-11-18 08-54-40

I agree with @sarah-clements, this looks pretty nice. Few polishes here and there and it should be ready to land.

**Comment by gregtatum** [comment]
Sorry, I missed the feedback here, as I was looking at the review status flag, and hadn't configured notifications for individual comments.

I switched teams from the Perf Tools team to the Internationalization team, so this is a bit of a low priority for me at the moment. I'll keep it in my backlog to finish up, so it's worth keeping this PR open, but I don't have a specific timeframe, as I still have a few things in my profiler backlog. If someone wanted to fix up the comments and merge it, I wouldn't be opposed, even if it's not as typical to do that with Mozilla patches. Otherwise I'll keep it in my backlog.

**Comment by jmaher** [comment]
@gregtatum I see your last comment about switching teams, did you want to finish this up?  anyone we could ask to finish it up?

**Pull request closed without merging**